File: Data Structures & Algorithms/island-perimeter/submission-1.cpp
class Solution {
public:
    set<pair<int, int>> visit = {};

    int dfs(vector<vector<int>>& grid, int i, int j) {
        if (i >= grid.size() or j >= grid[0].size() or i < 0 or j < 0 or grid[i][j] == 0)
            return 1;

        if (visit.count({i, j}))
            return 0;

        visit.insert({i, j});
        return (dfs(grid, i, j+1) + dfs(grid, i+1, j) + dfs(grid, i, j-1) + dfs(grid, i-1, j));
    }

    int islandPerimeter(vector<vector<int>>& grid) {
        for (int i = 0; i < grid.size(); i++)
            for (int j = 0; j < grid[0].size(); j++)
                if (grid[i][j] == 1)
                    return dfs(grid, i, j);
    }
};

File: Data Structures & Algorithms/count-number-of-islands/submission-1.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        # for a very large matrix, the recursion depth might hit. so we can use a stack based iterative approach.

        count = 0
        rows, cols = len(grid), len(grid[0])

        def dfs(i, j):
            stack = [(i, j)]
            grid[i][j] = "0"

            while stack:
                x, y = stack.pop()
                for dx, dy in [(1, 0), (0, 1), (-1, 0), (0, -1)]:
                    nx, ny = x + dx, y + dy
                    if 0 <= nx < rows and 0 <= ny < cols and grid[nx][ny] == "1":
                        grid[nx][ny] = "0"
                        stack.append((nx, ny))
        

        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == "1":
                    dfs(i, j)
                    count += 1
        
        return count